syntax = "proto3";

message Destination {
  enum Microcontroller {
    AUTO = 0;
    PICO_1 = 1;
    PICO_2 = 2;
    AVR = 3;
  };

  Microcontroller microcontroller = 1;
  string          part = 2;
  int32           baud = 3;
}

message Request {
  message FirmwareUpload {
    Destination destination = 1;
    bool        verify = 2;
    bool        verbose = 3;
    bool        payload_compressed = 4;
    bytes       payload = 99;
  }

  message AvrFuseProgramming {
    Destination  destination  = 1;
    uint32       low          = 2;
    uint32       high         = 3;
    bool         has_extended = 4;
    uint32       extended     = 5;
  }

  message Reset {
    uint32 time_ms = 1;
  }

  oneof request {
    FirmwareUpload     firmware_upload  = 1;
    Destination        test_connection  = 2;
    AvrFuseProgramming fuse_programming = 3;
    Reset              reset            = 4;
  }
}

//---------------------------------------

message Response {

  enum ResultCode { SUCCESS = 0; FAILURE = 1; ONGOING = 2; };

  message Result {
    ResultCode result_code = 1;
    string     errors   = 2;
    string     messages = 3;
  }

  oneof response {
    Result result      = 1;
    bytes  message     = 2;
  }

}